File: realtime/android/webrtc/app/src/test/java/ai/inworld/voiceagent/ServerEventDecodingTest.kt
package ai.inworld.voiceagent

import ai.inworld.voiceagent.realtime.events.ServerEvent
import org.junit.Assert.assertEquals
import org.junit.Test

class ServerEventDecodingTest {
    @Test
    fun outputTextDeltaVariants() {
        for (type in listOf(
            "response.output_text.delta",
            "response.text.delta",
            "response.audio_transcript.delta",
            "response.output_audio_transcript.delta",
        )) {
            val event = ServerEvent.decode("""{"type": "$type", "delta": "Hel"}""")
            assertEquals("for type $type", ServerEvent.OutputTextDelta("Hel"), event)
        }
    }

    @Test
    fun transcriptDone() {
        val event = ServerEvent.decode("""{"type": "response.output_audio_transcript.done", "transcript": "Hello there."}""")
        assertEquals(ServerEvent.TranscriptDone("Hello there."), event)
    }

    @Test
    fun contentPartDone() {
        val event = ServerEvent.decode("""{"type": "response.content_part.done", "part": {"transcript": "Hi!"}}""")
        assertEquals(ServerEvent.TranscriptDone("Hi!"), event)
    }

    @Test
    fun contentPartDoneWithoutTranscript() {
        val event = ServerEvent.decode("""{"type": "response.content_part.done", "part": {}}""")
        assertEquals(ServerEvent.TranscriptDone(null), event)
    }

    @Test
    fun inputTranscriptionDelta() {
        val event = ServerEvent.decode(
            """{"type": "conversation.item.input_audio_transcription.delta", "item_id": "i1", "delta": "what"}""",
        )
        assertEquals(ServerEvent.InputTranscriptionDelta("what"), event)
    }

    @Test
    fun inputTranscriptionCompleted() {
        val event = ServerEvent.decode(
            """{"type": "conversation.item.input_audio_transcription.completed", "transcript": "What time is it?"}""",
        )
        assertEquals(ServerEvent.InputTranscriptionCompleted("What time is it?"), event)
    }

    @Test
    fun speechStarted() {
        assertEquals(ServerEvent.SpeechStarted, ServerEvent.decode("""{"type": "input_audio_buffer.speech_started"}"""))
    }

    @Test
    fun outputItemAdded() {
        assertEquals(ServerEvent.OutputItemAdded, ServerEvent.decode("""{"type": "response.output_item.added", "item": {}}"""))
    }

    @Test
    fun responseDone() {
        assertEquals(
            ServerEvent.ResponseDone,
            ServerEvent.decode("""{"type": "response.done", "response": {"status": "completed"}}"""),
        )
    }

    @Test
    fun backchannelAudioDelta() {
        val event = ServerEvent.decode("""{"type": "response.backchannel.audio.delta", "backchannel_id": "b1", "delta": "AAEC"}""")
        assertEquals(ServerEvent.BackchannelAudioDelta("AAEC"), event)
    }

    @Test
    fun backchannelAudioDone() {
        val event = ServerEvent.decode("""{"type": "response.backchannel.audio.done", "backchannel_id": "b1", "phrase": "uh-huh"}""")
        assertEquals(ServerEvent.BackchannelAudioDone("uh-huh"), event)
    }

    @Test
    fun backchannelSkipped() {
        val event = ServerEvent.decode("""{"type": "response.backchannel.skipped", "backchannel_id": "b1", "reason": "deadline_missed"}""")
        assertEquals(ServerEvent.BackchannelSkipped("deadline_missed"), event)
    }

    @Test
    fun errorEvent() {
        val event = ServerEvent.decode("""{"type": "error", "error": {"message": "bad things"}}""")
        assertEquals(ServerEvent.Error("bad things"), event)
    }

    @Test
    fun errorEventTopLevelMessage() {
        val event = ServerEvent.decode("""{"type": "error", "message": "top-level"}""")
        assertEquals(ServerEvent.Error("top-level"), event)
    }

    @Test
    fun unknownTypeNeverThrows() {
        assertEquals(
            ServerEvent.Unknown("response.output_audio.delta"),
            ServerEvent.decode("""{"type": "response.output_audio.delta", "delta": "AAAA"}"""),
        )
    }

    @Test
    fun garbageInput() {
        assertEquals(ServerEvent.Unknown(""), ServerEvent.decode("not json"))
    }
}

File: realtime/flutter/webrtc/test/server_event_decoding_test.dart
import 'dart:convert';

import 'package:flutter_test/flutter_test.dart';
import 'package:inworld_voice_agent/realtime/events/server_event.dart';

ServerEvent decode(Map<String, dynamic> obj) => ServerEvent.decode(jsonEncode(obj));

void main() {
  test('output text delta variants map to one case', () {
    for (final type in [
      'response.output_text.delta',
      'response.text.delta',
      'response.audio_transcript.delta',
      'response.output_audio_transcript.delta',
    ]) {
      final event = decode({'type': type, 'delta': 'hi'});
      expect(event, isA<OutputTextDelta>());
      expect((event as OutputTextDelta).delta, 'hi');
    }
  });

  test('transcript done from top-level and content_part', () {
    expect(
      (decode({'type': 'response.output_audio_transcript.done', 'transcript': 'done'})
              as TranscriptDone)
          .text,
      'done',
    );
    expect(
      (decode({
        'type': 'response.content_part.done',
        'part': {'transcript': 'part'},
      }) as TranscriptDone)
          .text,
      'part',
    );
  });

  test('input transcription delta and completed', () {
    expect(
      (decode({
        'type': 'conversation.item.input_audio_transcription.delta',
        'delta': 'ab',
      }) as InputTranscriptionDelta)
          .delta,
      'ab',
    );
    expect(
      (decode({
        'type': 'conversation.item.input_audio_transcription.completed',
        'transcript': 'final',
      }) as InputTranscriptionCompleted)
          .transcript,
      'final',
    );
  });

  test('lifecycle events', () {
    expect(decode({'type': 'input_audio_buffer.speech_started'}), isA<SpeechStarted>());
    expect(decode({'type': 'response.output_item.added'}), isA<OutputItemAdded>());
    expect(decode({'type': 'response.done'}), isA<ResponseDone>());
  });

  test('backchannel events', () {
    expect(
      (decode({'type': 'response.backchannel.audio.delta', 'delta': 'AAA='})
              as BackchannelAudioDelta)
          .base64Pcm16,
      'AAA=',
    );
    expect(
      (decode({'type': 'response.backchannel.audio.done', 'phrase': 'uh-huh'})
              as BackchannelAudioDone)
          .phrase,
      'uh-huh',
    );
    expect(
      (decode({'type': 'response.backchannel.skipped', 'reason': 'too soon'})
              as BackchannelSkipped)
          .reason,
      'too soon',
    );
  });

  test('error from nested and flat message', () {
    expect(
      (decode({
        'type': 'error',
        'error': {'message': 'nested'},
      }) as ErrorEvent)
          .message,
      'nested',
    );
    expect((decode({'type': 'error', 'message': 'flat'}) as ErrorEvent).message, 'flat');
  });

  test('unknown and malformed fall back to UnknownEvent', () {
    expect((decode({'type': 'something.new'}) as UnknownEvent).type, 'something.new');
    expect(ServerEvent.decode('not json'), isA<UnknownEvent>());
    expect(ServerEvent.decode('{}'), isA<UnknownEvent>());
  });
}
